### PR TITLE
api: add /stake/pool/.../full/addrs endpoints

### PR DESCRIPTION
**Pull request by chappjc**
This adds new `/stake/pool/.../full/addrs` API endpoints for retrieving the
stakesubmission addresses for live tickets (at best block or the specified
block, depending on which endpoint is used).

### CODE REVIEW

**Comment by chappjc** [comment]
The idea is that you can gain insight on the stake pool.  e.g. `wget -qO- http://127.0.0.1:7777/api/stake/pool/b/300100/full/addrs | jq 'unique | map(select(test("^Ds"))), map(select(test("^Dc"))) | length'`

**Comment by chappjc** [comment]
Since this proposed endpoint is easily abused, it may need:

- The option to disable it.
- A mutex or try lock to prevent concurrent calls (a try lock that is not obtained could return an http status code of 503 and `Retry-After` response header set to a ~30 seconds delta).
- caching a map of ticket hashes to stake submission addresses (ticket addresses).  As of block 348921, there are 1,778,672 distinct ticket hashes.  Assuming ~32+24 = 56 bytes per hash-address pair, this is about 95 MiB for the entire set.  The size of cache can be limited, perhaps with a priority queue.  Also, since there are only 488,389 distinct ticket addresses, the ticket hash map could contain pointers to addresses (many to one).

**Comment by chappjc** [comment]
To prevent abuse, options 2 and 3 in the previous comment were implemented.

**Comment by chappjc** [comment]
On startup, the ChainDB constructor preloads the entire historical ticket hash-address map data for the `/stake/pool/.../full/addrs` API endpoints. As of block 350000, this is about 1.7 million ticket hashes (32 bytes each), and 500k addresses (~36 bytes each), for a total of less than 70 MiB.

File: README.md [hunk 1]
```diff
@@ -660,36 +670,37 @@ url query, a value of `true` will put all pool values and pool sizes into
...
-| --------------------------------- | --------------------- | --------------------------- |
-| The current agenda and its status | `/stake/vote/info`    | `dcrjson.GetVoteInfoResult` |
-| All agendas high level details    | `/agendas`            | `[]types.AgendasInfo`       |
-| Details for agenda {agendaid}     | `/agendas/{agendaid}` | `types.AgendaAPIResponse`   |
+| Votes and Agendas Info                            | Path                  | Type                         |
+| ------------------------------------------------- | --------------------- | ---------------------------- |
+| The current agenda and its status                 | `/stake/vote/info`    | `dcrjson.GetVoteInfoResult`  |
+| All agendas high level details                    | `/agendas`            | `[]types.AgendasInfo`        |
+| Details for agenda {agendaid}                     | `/agendas/{agendaid}` | `types.AgendaAPIResponse`    |
+| Vote information for a Politeia proposal {token}  | `/proposal/{token}`   | `dbtypes.ProposalChartsData` |
```
Comment (dmigwi) [inline README.md]: ~Vote information~ => Votes

**Comment by buck54321** [comment]
I'm seeing quite a bit more memory used than anticipated. > 500 MB over master.

**Comment by chappjc** [comment]
Is that MemStats.HeapInUse, or RSS from the OS?

**Comment by buck54321** [comment]
Resident size on Linux. Also quickly checked the Windows GUI and working set size.

**Comment by chappjc** [comment]
With go 1.12, OS reported usage is misleading, but that does seem like a large increase.  I too am seeing 1.1 GB (master) -> 1.6GB (PR) for RSS in the task manager.

In terms of [`MemStats`](https://golang.org/pkg/runtime/#MemStats):


|                    |                  master                 | PR (pre-GC)          | PR (post-GC)         |
|:------------------:|:----------------------------------------------:|-----------------------------------------------|-----------------------------------------------|
|         sys (VIRT)        |            680.12MB (713154808 bytes)          |  1.38GB (1482622200 bytes)                    |  1.38GB (1482622200 bytes)                    |
| lookups            | 0                                              | 0                                             | 0                                             |
| mallocs            | 87212642                                       | 118486178                                     | 118491967                                     |
| frees              | 82150203                                       | 106881590                                     | 116166039                                     |
| heap-alloc (includes eligible garbage)         |  588.97MB (617575296 bytes)                    |  896.16MB (939695536 bytes)                   |  674.76MB (707538224 bytes)                   |
| heap-sys (VIRT)          |  638.28MB (669286400 bytes)                    |  1.31GB (1407778816 bytes)                    |  1.31GB (1407811584 bytes)                    |
| heap-idle (ready to return to OS)         |  3.23MB (3383296 bytes)                        |  325.62MB (341442560 bytes)                   |  519.10MB (544317440 bytes)                   |
| **heap-in-use**        |  635.05MB (665903104 bytes)                    |  1016.94MB (1066336256 bytes)                 |  823.49MB (863494144 bytes)                   |
| heap-released      |  0 bytes                                       |  0 bytes                                      |  0 bytes                                      |
| heap-objects       | 5062439                                        | 11604588                                      | 2325928                                       |
| stack-in-use       |  1.72MB (1802240 bytes)                        |  1.44MB (1507328 bytes)                       |  1.41MB (1474560 bytes)                       |
| stack-sys  (VIRT)        |  1.72MB (1802240 bytes)                        |  1.44MB (1507328 bytes)                       |  1.41MB (1474560 bytes)                       |
| stack-mspan-inuse  |  7.33MB (7688304 bytes)                        |  9.00MB (9440352 bytes)                       |  6.57MB (6887232 bytes)                       |
| stack-mspan-sys    |  7.67MB (8044544 bytes)                        |  11.27MB (11812864 bytes)                     |  11.27MB (11812864 bytes)                     |
| stack-mcache-inuse |  54.25KB (55552 bytes)                         |  54.25KB (55552 bytes)                        |  54.25KB (55552 bytes)                        |
| stack-mcache-sys   |  64.00KB (65536 bytes)                         |  64.00KB (65536 bytes)                        |  64.00KB (65536 bytes)                        |
| other-sys          |  4.31MB (4522176 bytes)                        |  3.11MB (3262065 bytes)                       |  3.11MB (3262065 bytes)                       |
| gc-sys             |  26.58MB (27873280 bytes)                      |  53.99MB (56608768 bytes)                     |  53.99MB (56608768 bytes)                     |
| next-gc            |  when heap-alloc >= 789.50MB (827850560 bytes) |  when heap-alloc >= 1.32GB (1421881968 bytes) |  when heap-alloc >= 1.32GB (1415029920 bytes) |
| last-gc            |  2019-06-05 20:59:48.315165087 -0500 CDT       |  2019-06-05 21:08:24.662657444 -0500 CDT      |  2019-06-05 21:08:54.18010438 -0500 CDT       |
| gc-pause-total     |  3.202226ms                                    |  2.788284ms                                   |  2.831195ms                                   |
| gc-pause           | 32965                                          | 35147                                         | 42911                                         |
| num-gc             | 29                                             | 32                                            | 33                                            |

**Comment by chappjc** [comment]
(so this isn't behaving as intended, and definitely won't scale).  Gonna have a closer look.

**Comment by chappjc** [comment]
Rescoped for 5.1

**Comment by chappjc** [comment]
This is not a critical feature, and the solution needs better memory management.  Closing for now.

**Pull request closed without merging**